### PR TITLE
feat: smt data structs

### LINKED ISSUE #22: Replace some of the generated SMT strings by proper data structures

### PR DESCRIPTION
**Pull request by rkrasiuk**
close #22

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -0,0 +1,159 @@
+# This file is automatically @generated by Cargo.
```
Comment (chriseth) [inline Cargo.lock:1]: Please don't check in the cargo lockfile.

File: src/smt.rs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        Self::Indexed("BitVec".to_owned(), index)
+    }
+}
+
+pub enum SMTExpression {
+    Eq(String, String),
+    Not(String),
+    Assert(String),
+    DeclareConst(String, SMTIdentifier),
+    DefineConst(String, SMTIdentifier, String),
```
Comment (leonardoalt) [inline src/smt.rs:44]: I think we should separate these into SMTExpression which contains the operators and SMTLibCommand (which includes Assert, DeclareConst, DefineConst).

File: src/smt.rs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        Indexed(symbol, index) "(_ {} {})",
+    }
+);
+impl_smt!(
+    enum SMTLiteral {
+        Hex(val) "#x{}",
+    }
+);
+impl_smt!(
+    enum SMTExpression {
```
Comment (leonardoalt) [inline src/smt.rs:88]: Nice!

File: src/smt.rs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        self.to_owned()
+    }
+}
+
+pub enum SMTLiteral {
+    Hex(String),
+}
+
+impl SMTLiteral {
+    pub fn hex_zero() -> Self {
```
Comment (chriseth) [inline src/smt.rs:23]: The important aspect of this is that it's a BV constant with 256 bits - maybe we could call this `bv256_zero` or something like that?

File: src/smt.rs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+impl SMTIdentifier {
+    pub fn bitvec(index: usize) -> Self {
+        Self::Indexed("BitVec".to_owned(), index)
+    }
+}
+
+pub enum SMTExpression {
+    Eq(String, String),
+    Not(String),
+    Assert(String),
```
Comment (chriseth) [inline src/smt.rs:42]: Strictly, Assert, DeclareConst and DefineConst are not expressions but statements. Not sure if this distinction is too important, though.

File: src/smt.rs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+impl SMTIdentifier {
+    pub fn bitvec(index: usize) -> Self {
+        Self::Indexed("BitVec".to_owned(), index)
+    }
+}
+
+pub enum SMTExpression {
+    Eq(String, String),
```
Comment (chriseth) [inline src/smt.rs:40]: Wouldn't it be better to have the arguments here be SMTExpressions as well?

File: src/smt.rs [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+impl SMTLiteral {
+    pub fn hex_zero() -> Self {
+        Self::Hex("0".repeat(64))
+    }
+}
+
+pub enum SMTIdentifier {
+    Simple(String),
+    Indexed(String, usize), // for now supports only one index
```
Comment (leonardoalt) [inline src/smt.rs:30]: What would be other indices? This reminds me a bit of https://github.com/facebookincubator/smt2utils/blob/main/smt2parser/src/concrete.rs#L65
Wondering if we can get inspired by that module also

**Review by leonardoalt — COMMENTED** [review]
Nice, thanks for the PR! Left some comments, I think they should be easy to address/agree on.

File: src/encoder.rs [hunk 8]
```diff
@@ -89,7 +89,10 @@ impl<InstructionsType: Instructions> Encoder<InstructionsType> {
     pub fn encode_function(&mut self, function: &FunctionDefinition) -> FunctionVariables {
         for param in &function.parameters {
             let var = self.ssa_tracker.introduce_variable(param);
-            self.out(format!("(declare-const {} (_ BitVec 256))", var.name));
+            self.out(SMTExpression::DeclareConst(
+                var.name.as_smt(),
+                SMTIdentifier::bitvec(256),
```
Comment (chriseth) [inline src/encoder.rs:94]: At first, I was a bit confused why SMTIdentifier is used here - maybe you could make it clearer that this is a sort?

**Review by chriseth — CHANGES_REQUESTED** [review]
Thank you so much for your PR!
In general, it looks really good! I have some comments about the way it is structured.

File: src/smt.rs (hunk 7 above)
Comment (rkrasiuk) [reply src/smt.rs:30]: there was no use case for multiple rn, but the [lib ref](https://smtlib.cs.uiowa.edu/papers/smt-lib-reference-v2.6-r2017-07-18.pdf) stated that indexes can be repeated
Screenshot 2022-09-12 at 11 55 38
thanks for the lib link, i'll take a look, however, i think qualified identifiers are a bit different from regular ones

**Comment by rkrasiuk** [comment]
@leonardoalt @chriseth thanks so much for your comments, i've addressed most of them. @chriseth regarding `SMTExpression`, recursive types are a bit of mouthful in Rust & I've done it by introducing ` Box<dyn SMTFormat>`. if heap allocations are a big deal here, happy to hack a way around it by maybe introducing some top level SMT unit type

**Comment by chriseth** [comment]
Hm, I'm wondering, if we store the recursive parts as `dyn SMTFormat`, then all information we can extract out of it is the string-formatted representation.
I think we have to decide if we want to "do something" on these structures, or if this is all just helpers to construct smt string representations. If we do not want to run analyses or something on the representation, SMTExpressions could also just be something like
```
enum SExpression {
  Atom(String),
  SubExpression(Vec<SExpression>),
}
```

**Comment by leonardoalt** [comment]
@chriseth ideally, i think at some point we should have proper data structures that know nothing about formating or strings, like we have in the c++ Expression, but with better types. I think we don't need to do that here though to advance it, and keep changing in the future.

File: src/smt.rs (hunk 7 above)
Comment (leonardoalt) [reply src/smt.rs]: Yep, agree.

File: node_modules/.yarn-integrity [hunk 9]
```diff
@@ -0,0 +1,10 @@
+{
+  "systemParams": "darwin-arm64-93",
+  "modulesFolders": [],
+  "flags": [],
+  "linkedModules": [],
+  "topLevelPatterns": [],
+  "lockfileEntries": {},
+  "files": [],
+  "artifacts": {}
+}
```
Comment (leonardoalt) [inline node_modules/.yarn-integrity:10]: I think this file went in by mistake?

**Comment by leonardoalt** [comment]
@rkrasiuk could you please just remove the yarn file and squash? Then I'd say we should merge and continue from there.

**Comment by rkrasiuk** [comment]
@leonardoalt apologies, that was probably my erratic behavior of switching between open tabs too fast and hitting keystrokes. done ✔️

**Review by leonardoalt — APPROVED** [review]
Awesome, thanks again!

**Pull request merged**